### PR TITLE
[BEAM-5504] Introduce PubsubAvroTable

### PR DESCRIPTION
**Pull request by milantracy**
The PR introcduces AVRO support for PubsubTable.

The initial attempt is to make avro format a parity with the existing json format.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by amaliujia** [comment]
Nice! Please let me know when this PR is ready to be reviewed. 

cc @kennknowles

**Comment by amaliujia** [comment]
R: @amaliujia

**Comment by amaliujia** [comment]
Run SQL PostCommit

**Comment by milantracy** [comment]
thanks @amaliujia , i believe the pr is not ready for review yet.
I will make some changes.

**Comment by amaliujia** [comment]
Run SQL PostCommit

**Comment by amaliujia** [comment]
Run SQL PostCommit

**Comment by milantracy** [comment]
retest this please

**Comment by milantracy** [comment]
Run SQL PostCommit

**Comment by milantracy** [comment]
@amaliujia do you want to take a look? Feel free to tag anyone who is free for a code review

**Comment by milantracy** [comment]
Run SQL PostCommit

**Comment by amaliujia** [comment]
Thanks @milantracy! I will take a look!

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/pubsub/AvroPubsubMessageToRow.java [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+          return payload;
+        default:
+          throw new IllegalArgumentException(
+              "Unexpected field '"
+                  + field.getName()
+                  + "' in top level schema"
+                  + " for Pubsub message. Top level schema should only contain "
+                  + "'timestamp', 'attributes', and 'payload' fields");
+      }
+    }
```
[the remark this answers is not in the document]
Comment (milantracy) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/pubsub/AvroPubsubMessageToRow.java]: I would love to remove `nested schema` for the table provider here.
I don't think IO module is a good place to read/write a specific attribute, which might be wrong.

Meanwhile, I am curious whether or not there would be a good place to let users know the `nested schema` is being deprecated.

**Comment by milantracy** [comment]
> Thanks for this @milantracy! Right now my only concern with this is that it's duplicating a lot of code from the PubsubJsonTableProvider for supporting both flat and nested schemas. What do you think about adding an abstract `PubsubTableProvider` that includes the logic for the different types of schemas, and then each implementation can be responsible for translating the payload to/from avro or JSON?
> 
> That way when we get rid of support for nested schemas we only have to do it in one place.

@TheNeuralBit Thanks, this is a good idea, I will created a abstract class `PubsubTableProvider` and two final class `JsonPubsubTableProvider` and `AvroPubsubTableProvider`

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/pubsub/AvroPubsubMessageToRow.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (TheNeuralBit) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/pubsub/AvroPubsubMessageToRow.java]: Good point. I brought this up on the [mailing list](https://lists.apache.org/thread.html/rb83f154846a3a38eae30b667cee52d6b8f0e94769e71b2b7516cb2ce%40%3Cdev.beam.apache.org%3E). If there aren't any objections there I think we should add a warning message when people use nested schemas.

**Comment by TheNeuralBit** [comment]
@milantracy have you had a chance to work on that change?

**Comment by milantracy** [comment]
@TheNeuralBit working on it, but need a few more time.

Is anything you are working on blocked?

**Comment by TheNeuralBit** [comment]
Nope, not blocked! I just wanted to check in in case there's anything I can do to help :)

**Comment by amaliujia** [comment]
Sorry I was busy on something else so didn't review this PR in time. But great thanks @TheNeuralBit jumps in to review it!

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/utils/AvroUtils.java [hunk 2]
```diff
@@ -339,6 +350,35 @@ public static Schema toBeamSchema(org.apache.avro.Schema schema) {
...
+      GenericRecord record = datumReader.read(null, decoder);
+      if (record != null) {
+        return record;
+      }
+    } catch (IOException exception) {
+      throw new RuntimeException("Failed to extract the record from the payload.", exception);
+    }
+    throw new RuntimeException("No record is extracted from the payload");
+  }
+
```
[the remark this answers is not in the document]
Comment (amaliujia) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/utils/AvroUtils.java]: I think `AvroUtils.java` is good place. Coder seems not being used as util class.

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/pubsub/AvroPubsubMessageToRow.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (amaliujia) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/pubsub/AvroPubsubMessageToRow.java]: If `nested schema` will be deprecated, +1 to not just support it for `AvroPubsubMessageToRow`

**Comment by aaltay** [comment]
Is this PR still active?

**Comment by milantracy** [comment]
@aaltay i think it is still alive, i will update the pr

**Comment by TheNeuralBit** [comment]
@milantracy do you have time to finish this up? I understand its likely priorities have changed over the last few months

**Comment by milantracy** [comment]
Thanks @TheNeuralBit I will update the PR at the weekend probably, SGTY?

**Comment by TheNeuralBit** [comment]
Yep SGTM :)

**Comment by aaltay** [comment]
@milantracy - How is this PR going?

**Comment by TheNeuralBit** [comment]
FYI @sclukas77 is working on https://github.com/apache/beam/pull/12090 which will conflict with this when its merged.

It's mostly just moving code around (from extensions/sql -> core) so the conflicts won't be too difficult to resolve, but they won't be trivial.

**Comment by milantracy** [comment]
@TheNeuralBit Thanks for the heads up
@aaltay thanks for asking, i am seeing a couple of unit test failure, as long as they got fixed, i will update the PR :)

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/utils/AvroUtils.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (milantracy) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/utils/AvroUtils.java:382]: I may keep the util functions here as I don't see a way to convert between GenericRecord and Row via AvroCoder

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/pubsub/AvroPubsubMessageToRow.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (milantracy) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/pubsub/AvroPubsubMessageToRow.java]: FYI,
I dump all nested schema related stuff for avro support.

If a user tries to use nested schema avro message, it throws error during the table initilizaiton.

For json, nested schema support is left as it was

**Comment by milantracy** [comment]
@TheNeuralBit @amaliujia  @aaltay sorry for updating the PR after such a long time , feel free to leave your comments.

**Comment by milantracy** [comment]
seems that PreCommit is not related to the PR

https://ci-beam.apache.org/job/beam_PreCommit_Java_Commit/12039/console

**Comment by milantracy** [comment]
Run Java PreCommit

**Comment by TheNeuralBit** [comment]
Hey @milantracy sorry for the churn on this... I didn't realize this was ready to go and I went ahead and merged #12090 which is causing the conflicts.

That PR moves the majority of the logic for PubsubJsonTableProvider next to the GCP IO module. Some of it was just straight file renames (e.g. PubsubMessageToRow) which shouldn't be too tough to resolve but PubsubIOJsonTable and PubsubJsonTableProvider are trickier, as we added a new interface that will be amenable for use in cross-language (see s.apache.org/beam-schema-io for details on this)

If you don't have bandwidth to resovle the conflicts, I could probably help out after next week.

**Comment by milantracy** [comment]
@TheNeuralBit thanks, I have been aware of changes by #12090 and looking into it.

I am working on this PR accordingly.

I agree that IOTables and TableProviders look much different from the previous version.

**Pull request closed without merging**